src/coreclr/jit/gentree.cpp

```diff
@@ -2155,16 +2155,27 @@ GenTree* Compiler::getArrayLengthFromAllocation(GenTree* tree DEBUGARG(BasicBloc
 
         if (call->gtCallType == CT_HELPER)
         {
-            switch (eeGetHelperNum(call->gtCallMethHnd))
+            CorInfoHelpFunc helper = eeGetHelperNum(call->gtCallMethHnd);
+            switch (helper)
             {
+                case CORINFO_HELP_NEWARR_1_FROZEN:
                 case CORINFO_HELP_NEWARR_1_DIRECT:
                 case CORINFO_HELP_NEWARR_1_OBJ:
                 case CORINFO_HELP_NEWARR_1_VC:
                 case CORINFO_HELP_NEWARR_1_ALIGN8:
                 {
-                    // This is an array allocation site. Grab the array length node.
-                    arrayLength = call->gtArgs.GetArgByIndex(1)->GetNode();
-                    break;
+                    if (opts.IsReadyToRun() && helper == CORINFO_HELP_NEWARR_1_FROZEN)
+                    {
+                        // CORINFO_HELP_NEWARR_1_FROZEN is used in both JIT and R2R,
+                        // for R2R take the same path as CORINFO_HELP_READYTORUN_NEWARR_1
```

Comment:
Why do we want to do that? `CORINFO_HELP_NEWARR_1_FROZEN` has same signature for both JIT and R2R

Reply:
I've simplifed the whole thing with just `GetUserArgByIndex` that ignores all "non-user" args